ruby/ql/lib/codeql/ruby/ast/internal/Constant.qll

```diff
@@ -0,0 +1,32 @@
+cached
+newtype TConstantValue =
+  TInt(int i) { any(RequiredConstantValue x).requiredInt(i) } or
+  TFloat(float f) { any(RequiredConstantValue x).requiredFloat(f) } or
+  TRational(int numerator, int denominator) {
+    any(RequiredConstantValue x).requiredRational(numerator, denominator)
+  } or
+  TComplex(float real, float imaginary) {
+    any(RequiredConstantValue x).requiredComplex(real, imaginary)
+  } or
+  TString(string s) { any(RequiredConstantValue x).requiredString(s) } or
+  TSymbol(string s) { any(RequiredConstantValue x).requiredSymbol(s) } or
+  TBoolean(boolean b) { b in [false, true] } or
+  TNil()
+
+private newtype TRequiredConstantValue = MkRequiredConstantValue()
+
+class RequiredConstantValue extends MkRequiredConstantValue {
```

Comment:
I don't really understand the meaning of "required" in this name. This seems to be a class that holds a set of predicates defining all constant values, but I'm not sure what is "required" about them. Maybe doc comment here would help clarify?

Reply:
I'll add ql doc.